vault/resource_cert_auth_backend_role.go

```diff
@@ -0,0 +1,264 @@
+
+	"github.com/hashicorp/vault/api"
+)
+
+func certAuthBackendRoleResource() *schema.Resource {
+	return &schema.Resource{
+		SchemaVersion: 1,
+
+		Create: certAuthResourceWrite,
+		Update: certAuthResourceUpdate,
+		Read:   certAuthResourceRead,
+		Delete: certAuthResourceRead,
+
+		Schema: map[string]*schema.Schema{
+			"name": &schema.Schema{
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+			"certificate": &schema.Schema{
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+			"allowed_names": &schema.Schema{
+				Type: schema.TypeSet,
+				Elem: &schema.Schema{
+					Type: schema.TypeString,
+				},
+				Optional: true,
+				Computed: true,
+			},
+			"required_extensions": &schema.Schema{
+				Type: schema.TypeSet,
+				Elem: &schema.Schema{
+					Type: schema.TypeString,
+				},
+				Optional: true,
+				Computed: true,
+			},
+			"ttl": &schema.Schema{
+				Type:     schema.TypeString,
+				Optional: true,
+				Computed: true,
+			},
+			"max_ttl": &schema.Schema{
+				Type:     schema.TypeString,
+				Optional: true,
+				Computed: true,
+			},
+			"period": &schema.Schema{
+				Type:     schema.TypeString,
+				Optional: true,
+				Computed: true,
+			},
+			"policies": &schema.Schema{
+				Type: schema.TypeSet,
+				Elem: &schema.Schema{
+					Type: schema.TypeString,
+				},
+				Optional: true,
+				Computed: true,
+			},
+			"display_name": &schema.Schema{
+				Type:     schema.TypeString,
+				Optional: true,
+				Computed: true,
+			},
+			"backend": &schema.Schema{
+				Type:     schema.TypeString,
+				Optional: true,
+				ForceNew: true,
+				Default:  "cert",
+				StateFunc: func(v interface{}) string {
+					return strings.Trim(v.(string), "/")
+				},
+			},
+		},
+	}
+}
+
+func certCertResourcePath(backend, name string) string {
+	return "auth/" + strings.Trim(backend, "/") + "/certs/" + strings.Trim(name, "/")
+}
+
+func certAuthResourceWrite(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*api.Client)
+
+	backend := d.Get("backend").(string)
+	name := d.Get("name").(string)
+
+	path := certCertResourcePath(backend, name)
+
+	data := map[string]interface{}{}
+	data["certificate"] = d.Get("certificate")
+
+	if v, ok := d.GetOk("allowed_names"); ok {
+		data["allowed_names"] = v.(*schema.Set).List()
+	}
+
+	if v, ok := d.GetOk("required_extensions"); ok {
+		data["required_extensions"] = v.(*schema.Set).List()
+	}
+
+	if v, ok := d.GetOk("ttl"); ok {
+		data["ttl"] = v.(string)
+	}
+
+	if v, ok := d.GetOk("max_ttl"); ok {
+		data["max_ttl"] = v.(string)
+	}
+
+	if v, ok := d.GetOk("period"); ok {
+		data["period"] = v.(string)
+	}
+
+	if v, ok := d.GetOk("policies"); ok {
+		data["policies"] = v.(*schema.Set).List()
+	}
+
+	if v, ok := d.GetOk("display_name"); ok {
+		data["display_name"] = v.(string)
+	}
+
+	log.Printf("[DEBUG] Writing %q to cert auth backend", path)
+	_, err := client.Logical().Write(path, data)
+	d.SetId(path)
```

Comment:
Could this be moved up a line so it's not between the err and the err check?

Reply:
No problem, will fix.